ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/Util/HardFork/Future.hs

```diff
@@ -0,0 +1,151 @@
+futureFirstEpochSize :: Future -> EpochSize
+futureFirstEpochSize future = case future of
+    NextEra  _slotLength epochSize _eraSize _future -> epochSize
+    FinalEra _slotLength epochSize                  -> epochSize
+
+-- | Future as of a slot within an epoch
+data SlotFuture =
+      InFinalEra SlotLength
+    | InNextEra  SlotLength NumSlots Future
+      -- ^ INVARIANT: @'NumSlots' > 0@
+
+-- | Length of each slot in the whole 'Future'
+futureSlotLengths :: Future -> Stream SlotLength
+futureSlotLengths = \case
+    FinalEra slotLength _epochSize ->
+        let x = slotLength :< x in x
+    NextEra slotLength epochSize eraSize future ->
+        nTimes (slotLength :<) eraSlots $
+        futureSlotLengths future
+      where
+        NumSlots eraSlots = calcEraSlots epochSize eraSize
+
+-- | @(slot, time left in slot, length of slot)@
+futureTimeToSlot :: Future
+                 -> NominalDiffTime
+                 -> (SlotNo, NominalDiffTime, SlotLength)
+futureTimeToSlot = \future d -> go2 0 d future
+  where
+    go acc d (InFinalEra slotLength) =
+        (SlotNo $ acc + n, timeInSlot, slotLength)
+      where
+        n = divide d slotLength
+        timeInSlot = d - multiply n slotLength
+    go acc d (InNextEra slotLength (NumSlots leftovers) future) =
+        case d `safeSub` remaining of
+          Nothing -> go  acc               d  (InFinalEra slotLength)
+          Just d' -> go2 (acc + leftovers) d' future
+      where
+        remaining = multiply leftovers slotLength
+
+    go2 acc d (FinalEra slotLength _epochSize) =
+        go acc d $ InFinalEra slotLength
+    go2 acc d (NextEra slotLength epochSize eraSize future) =
+        go acc d $ InNextEra slotLength (NumSlots eraSlots) future
+      where
+        NumSlots eraSlots = calcEraSlots epochSize eraSize
+
+-- | Which epoch the slot is in
+futureSlotToEpoch :: Future
+                  -> SlotNo
+                  -> EpochNo
+futureSlotToEpoch = \future (SlotNo s) -> EpochNo $ go 0 s future
+  where
+    go acc s = \case
+      FinalEra _slotLength (EpochSize epSz) ->
+          acc + s `div` epSz
+      NextEra _slotLength epochSize eraSize future ->
+          case s `safeSub` eraSlots of
+            Nothing -> go acc s (FinalEra _slotLength epochSize)
+            Just s' -> go (acc + n) s' future
+        where
+          EraSize n = eraSize
+          NumSlots eraSlots = calcEraSlots epochSize eraSize
+
+-- | Whether the epoch is in the first era
+futureEpochInFirstEra :: Future -> EpochNo -> Bool
+futureEpochInFirstEra = \case
+    NextEra _slotLength _epochSize (EraSize n) _future ->
+        \(EpochNo e) -> e < n
+    FinalEra{} -> const True
+
+{-------------------------------------------------------------------------------
+  Miscellany
+-------------------------------------------------------------------------------}
+
+divide :: NominalDiffTime -> SlotLength -> Word64
+divide d slotLength = floor $ d / getSlotLength slotLength
```

Comment:
Could use `Data.Fixed.div'`.

Reply:
Thanks, I've never internalized that module.